lib/registry/src/graphql/mod.rs

```diff
@@ -0,0 +1,166 @@
+pub(crate) mod mutations;
+pub(crate) mod proxy;
+pub(crate) mod queries;
+
+use graphql_client::*;
+use reqwest::{
+    blocking::{multipart::Form, Client},
+    header::USER_AGENT,
+};
+use std::env;
+use std::time::Duration;
+
+use crate::config::WasmerConfig;
+
+pub fn whoami_distro() -> String {
+    #[cfg(target_os = "wasi")]
+    {
+        whoami::os().to_lowercase()
+    }
+    #[cfg(not(target_os = "wasi"))]
+    {
+        whoami::distro().to_lowercase()
+    }
+}
+
+fn setup_client() -> Result<Client, anyhow::Error> {
+    let builder = Client::builder();
+    let builder = proxy::maybe_set_up_proxy_blocking(builder)?;
+    builder.build().map_err(|e| e.into())
+}
+
+/// This function is being used to "ping" the registry
+/// (see test_if_registry_present and see whether the response
+/// is valid JSON, it doesn't check the response itself,
+/// since the response format might change
+pub fn execute_query_modifier_inner_check_json<V, F>(
+    registry_url: &str,
+    login_token: &str,
+    query: &QueryBody<V>,
+    timeout: Option<Duration>,
+    form_modifier: F,
+) -> anyhow::Result<()>
+where
+    V: serde::Serialize,
+    F: FnOnce(Form) -> Form,
+{
+    let client = setup_client()?;
+
+    let vars = serde_json::to_string(&query.variables).unwrap();
+
+    let form = Form::new()
+        .text("query", query.query.to_string())
+        .text("operationName", query.operation_name.to_string())
+        .text("variables", vars);
+
+    let form = form_modifier(form);
+
+    let user_agent = format!(
+        "wasmer/{} {} {}",
+        env!("CARGO_PKG_VERSION"),
+        whoami::platform(),
+        whoami_distro(),
+    );
+
+    let mut res = client
+        .post(registry_url)
+        .multipart(form)
+        .bearer_auth(
+            env::var(WasmerConfig::ENV_VAR_WASMER_REGISTRY_TOKEN)
+                .ok()
+                .or_else(|| env::var(WasmerConfig::ENV_VAR_WASMER_REGISTRY_TOKEN_LEGACY).ok())
+                .unwrap_or_else(|| login_token.to_string()),
+        )
+        .header(USER_AGENT, user_agent);
+
+    if let Some(t) = timeout {
+        res = res.timeout(t);
+    }
+
+    let res = res.send()?;
+
+    let _: Response<serde_json::Value> = res.json()?;
+
+    Ok(())
+}
+
+pub fn execute_query_modifier_inner<R, V, F>(
+    registry_url: &str,
+    login_token: &str,
+    query: &QueryBody<V>,
+    timeout: Option<Duration>,
+    form_modifier: F,
```

Comment:
Nice I like this

Reply:
That's actually old code, I just moved it from graphql.rs to graphql/mod.rs.

There is a much cleaner way to execute the `graphql-client` queries, no need for form data.
The new execution logic is in client.rs ( fn execute(), fn execute_checked() )